[Take 2] Migrate library to use Halogen Hooks

JordanMartinez opened the pull request:
Uses `halogen-hooks` `v0.4.0` and `halogen-hooks-extra` `v0.6.0`.

See examples files for what the end interface looks like.

JordanMartinez committed:
Ignore Spago files

JordanMartinez committed:
Migrate project to Spago

JordanMartinez committed:
Convert library into a hook

JordanMartinez committed:
Add a dhall file for the examples folder

JordanMartinez committed:
Update examples to use hook version

JordanMartinez commented:
Once a new package set with `v0.5.0` of halogen-hooks-extra comes out, I think this PR can be merged. Right now, I'm overriding that package in the set.

JordanMartinez committed:
Update halogen-hooks to v0.3.0 & halogen-hooks-extra to v0.5.0

JordanMartinez committed:
Update PR to account for dependency changes

JordanMartinez committed:
Update examples to account for dependency updates

thomashoneyman commented on src/Select.purs:
```diff
@@ -90,266 +78,317 @@ derive instance ordVisibility :: Ord Visibility
-  , fiber :: Fiber Unit
-  }
+type SelectReturn
+      m toggleProps itemProps containerProps inputProps =
```
These don't feel like they'd be necessary in the return type. Can these type variables be pushed down into the props functions themselves?

```purs
toggleProps :: forall props. Array (HP.IProp (ToggleProps props) (HookM m Unit))
```

The monad type `m` would need to remain in the return type, though, to ensure all the `HookM m ...` unify.

I'm assuming you didn't want to put these in the type in the first place and ran into some type errors. If so, what happens if you `newtype` the Select return type? I just have the sense that this shouldn't be necessary.

JordanMartinez replied:
Yeah... I ran into "`slot` type escaped its scope" errors. Now that the `slot` type is no longer a part of the `HookM` definition, I might be able to make this change.

JordanMartinez replied:
Nope. It still happens. 
```
[1/1 EscapedSkolem] src/Select.purs:173:6

            v
  173    in Hooks.wrap Hooks.do
  174      state /\ modifyState <- useState
  175        { search: initialSearchValue
       ...
  205        , containerProps
  206        , inputProps: inputProps getState modifyState searchDebouncer inputRec.pushVisibilityChanged inputRec.pushSelectedIdxChanged
  207        }
             ^
  
  The type variable toggleProps, bound at
  
    src/Select.purs:106:20 - 106:97 (line 106, column 20 - line 106, column 97)
  
  has escaped its scope, appearing in the type
  
    Hooked m0 hooks1
      (UseDebouncer String
         (UseGet
            { highlightedIndex :: ...
            , search :: String
            , visibility :: Visibility
            }
            (UseState ... hooks1)
         )
      )
      { clearSearch :: HookM m0 Unit
      , containerProps :: Array (IProp ... ...)
      , highlightedIndex :: Maybe Int
      , inputProps :: Array (IProp ... ...)
      , itemProps :: Int -> Array ...
      , search :: String
      , setFocus :: Boolean -> HookM m0 Unit
      , setVisibility :: Visibility -> HookM m0 Unit
      , toggleProps :: Array (IProp ... ...)
      , visibility :: Visibility
      }
  
  in the expression ((bind ixBindIndexed) (useState { search: ...
                                                    , visibility: ...
                                                    , highlightedIndex: ...
                                                    }
                                          )
                    )
                    (\$5 ->
                       case $5 of
                         (Tuple state modifyState) -> ...
                    )
  in value declaration useSelect
```

thomashoneyman replied:
Hmm. This still strikes me as odd. This has the record printed out -- did you try adding a `newtype` here? I really don't want to force folks who want to annotate the return type to have to specify what props these functions would receive as input.

JordanMartinez replied:
I don't get what you mean by "adding a newtype here." So, I haven't tried doing that yet.

thomashoneyman replied:
Sorry, I mean `newtype SelectReturn m = SelectReturn { ... }`

JordanMartinez replied:
Done!

thomashoneyman commented on src/Select.purs:
```diff
@@ -90,266 +78,317 @@ derive instance ordVisibility :: Ord Visibility
+selectInput = do
+  let
+    ignoreEvent :: forall a. a -> HookM m Unit
+    ignoreEvent = \_ -> pure unit
```
This can also be defined as `mempty`, which simplifies your usage below:

```purs
, pushNewSearch: mempty
, pushVisibilityChanged: mempty
, ...
```

JordanMartinez replied:
Good point. I'll update that.

JordanMartinez replied:
Done in latest commits.

thomashoneyman commented on src/Select.purs:
```diff
@@ -90,266 +78,317 @@ derive instance ordVisibility :: Ord Visibility
-component
-  :: forall st query action slots input msg m
+useSelect
+  :: forall m toggleProps itemProps containerProps inputProps
```
If you're able to remove these from the return type then they can be removed here, too.

JordanMartinez replied:
Yup, the newtype fixed it. See latest commits.

thomashoneyman commented on src/Select.purs:
```diff
@@ -90,266 +78,317 @@ derive instance ordVisibility :: Ord Visibility
+      , visibility: state.visibility
+      , highlightedIndex: state.highlightedIndex
+
+      -- actions
```
Since there are only three actions being returned to the user, and the others are being used to set props, it seems like some refactoring could be done to reduce all the argument passing happening here.

Are you open to moving functions that depend on `state` up into let bindings here, to skip the argument passing, while leaving others which don't depend on it directly (like `setFocus`) in the where clause?

As a rule of thumb, anything that would take `state` or `modifyState` as an argument can be moved up into a let binding, and everything else can remain in the  `where` clause.

That way, once you've defined the actions that depend on state in `let`s you can just pass the resulting `HookM m Unit` functions directly to things like the `toggleProps` function, which reduces all the coordination being done among these various functions.

Still others can just be defined inline -- for example, `clearSearch` might as well just be defined inline as `clearSearch: modifyState _ { search = "" }` 

So, for example:

```purs
let
  setVisibility :: Visibility -> HookM m Unit
  setVisibility v = do
    when (state.visibility /= v) do
      modifyState (_ { visibility = v, highlightedIndex = Just 0 })
      inputRec.pushVisibilityChanged v

Hooks.pure
  { setVisibility
  , clearSearch: modifyState _ { search = "" }
  , toggleProps: toggleProps setVisibility toggleClick key
  }
where
toggleProps :: (Visibility -> HookM m Unit) -> ...
```

JordanMartinez replied:
I've inlined `clearSearch` in latest commits.

JordanMartinez replied:
I haven't done the other work yet because when I was thinking about this one, I [wondered whether `useGet` is necessary in case one wants to use a returned action in an Effect cleanup / forked computation](https://github.com/thomashoneyman/purescript-halogen-hooks/issues/30#issuecomment-627712157)

thomashoneyman replied:
It's not necessary if you're modifying state (`modifyState` always uses fresh state), but it would be if you are relying on the `state` variable to say perform an effect in one of these actions. But then you have to basically guess whether it would ever make sense to do a particular action in an effect cleanup / forked computation. If the answer is yes, and you do rely on the `state` variable to perform an effect, then you should use `useGet`.

JordanMartinez replied:
> It's not necessary if you're modifying state (`modifyState` always uses fresh state), but it would be if you are relying on the state variable to say perform an effect in one of these actions.

Another situation that can arise is if you need to refer to the state that a library returns. For example, if I had a hook like this:
```purescript
state' <- do
  state /\ stateId <- useState 0
  useTickEffect { } $ Hooks.modify_ stateId (_ + 1)
  pure state
useTickEffect { } do
  pure $ Just do
    liftEffect $ log $ "This value is stale: " <> show state
```
... the state value is stale. There are two ways this could be fixed
1. the library **user** uses `getState <- useGet state`

```purescript
state <- do
  state /\ stateId <- useState 0
  useTickEffect { } $ Hooks.modify_ stateId (_ + 1)
  pure state
getState <- useGet state
useTickEffect { } do
  pure $ Just do
    state' <- getState
    liftEffect $ log $ "This value is stale: " <> show state
    liftEffect $ log $ "This value is not stale: <> show state'
```

2. the library **author** returns `{ state, getState: Hooks.get stateId }`
```purescript
{ state, getState } <- do
  state /\ stateId <- useState 0
  useTickEffect { } $ Hooks.modify_ stateId (_ + 1)
  pure { state, getState: Hooks.get stateId }
useTickEffect { } do
  pure $ Just do
    state' <- getState
    liftEffect $ log $ "This value is stale: " <> show state
    liftEffect $ log $ "This value is not stale: <> show state'
```

thomashoneyman replied:
With Hooks 0.4.0 this is back to the usual.

JordanMartinez replied:
I'll update this PR to use that version later on today.

thomashoneyman commented on src/Select.purs:
```diff
@@ -90,266 +78,317 @@ derive instance ordVisibility :: Ord Visibility
+        -> (Visibility -> HookM m Unit)
+        -> Visibility
+        -> HookM m Unit
+      setVisibility getState modifyState pushVisibilityChanged v = do
```
By moving these implementations into `let` you can skip most of these function arguments and avoid using `getState` altogether.

JordanMartinez replied:
I decided not to do that. The number of args passed to the function are smaller now because of the `inputRec` being in scope. I did try putting these back into the `let` statement, but it still seemed verbose.

thomashoneyman commented on src/Select.purs:
```diff
@@ -90,266 +78,317 @@ derive instance ordVisibility :: Ord Visibility
+        -> (Int -> HookM m Unit)
+        -> KeyboardEvent
+        -> HookM m Unit
+      key getState modifyState pushVisibilityChanged pushSelectedIdxChanged ev = do
```
Even if you don't decide to move any of these into `let` bindings you can still refer to the input directly, without passing it as an argument. For example, both of these `push*` functions are already available via `inputRec` which is in scope.

JordanMartinez replied:
Oh, good point! I forgot about that.

JordanMartinez replied:
Done in latest commits

JordanMartinez committed:
Change `ignoreEvent` to `mempty`

JordanMartinez committed:
Inline clearSearch rather than defining it in where clause

JordanMartinez committed:
Get push* functions from inputRec in scope; don't pass it explicitly

thomashoneyman commented on spago.dhall:
```diff
@@ -0,0 +1,19 @@
+  ]
+, packages = ./packages.dhall
+, sources = [ "src/**/*.purs", "test/**/*.purs" ]
+}
```
If you'd like to switch the project to Spago, I'd like to ensure that generating a new `bower.json` file from the Spago dependencies produces a Bower file that matches the dependencies and versions currently there. That probably means that you move the test sources to a separate `test.dhall` file, and don't include dependencies like `debug`, `console`, `argonaut`, and `psci-support` (perhaps those should go in an `examples.dhall` file?) The `halogen-hooks` and `halogen-hooks-extra` are OK.

You might want to look at the migration to Spago for Halogen as a comparison:
https://github.com/purescript-halogen/purescript-halogen/pull/653

JordanMartinez replied:
How do I generate a bower.json file using Spago? It looks like `spago publish` is no longer a command in Spago `v0.15.2`.

Also, there is no `test` folder.

I've split up the dependencies between `spago.dhall` and `examples.dhall`

thomashoneyman replied:
That’s odd — I used it in the Halogen migration and I can’t find anything about it in the release notes. Without that command I don’t know how to generate one.

JordanMartinez replied:
D'oh! Looks like its `spago bump-version` that generates the `bower.json` file.

JordanMartinez replied:
Looks like it only worked for you because `halogen` was already on the Bower registry before they stopped all future uploads of new packages. Since our packages weren't already on there, this fails:
```
$ spago bump-version major
[info] Generating a new Bower config using the package set versions..
[error] Failed to run: `bower info --json purescript-halogen-hooks-extra#v0.6.0`
[error] [{
  "code": "ENOTFOUND",
  "data": {
    "endpoint": {
      "name": "",
      "source": "purescript-halogen-hooks-extra",
      "target": "v0.6.0"
    }
  },
  "id": "ENOTFOUND",
  "level": "error",
  "message": "Package purescript-halogen-hooks-extra not found",
  "stacktrace": "Error: Package purescript-halogen-hooks-extra not found\n    at createError (/home/jordan/.nvm/versions/node/v13.12.0/lib/node_modules/bower/lib/util/createError.js:4:15)\n    at /home/jordan/.nvm/versions/node/v13.12.0/lib/node_modules/bower/lib/core/resolverFactory.js:243:23\n    at _fulfilled (/home/jordan/.nvm/versions/node/v13.12.0/lib/node_modules/bower/lib/node_modules/q/q.js:854:54)\n    at /home/jordan/.nvm/versions/node/v13.12.0/lib/node_modules/bower/lib/node_modules/q/q.js:883:30\n    at Promise.promise.promiseDispatch (/home/jordan/.nvm/versions/node/v13.12.0/lib/node_modules/bower/lib/node_modules/q/q.js:816:13)\n    at /home/jordan/.nvm/versions/node/v13.12.0/lib/node_modules/bower/lib/node_modules/q/q.js:624:44\n    at runSingle (/home/jordan/.nvm/versions/node/v13.12.0/lib/node_modules/bower/lib/node_modules/q/q.js:137:13)\n    at flush (/home/jordan/.nvm/versions/node/v13.12.0/lib/node_modules/bower/lib/node_modules/q/q.js:125:13)\n    at processTicksAndRejections (internal/process/task_queues.js:79:11)"
}]

[error] Failed to run: `bower info --json purescript-halogen-hooks#v0.4.0`
[error] [{
  "code": "ENOTFOUND",
  "data": {
    "endpoint": {
      "name": "",
      "source": "purescript-halogen-hooks",
      "target": "v0.4.0"
    }
  },
  "id": "ENOTFOUND",
  "level": "error",
  "message": "Package purescript-halogen-hooks not found",
  "stacktrace": "Error: Package purescript-halogen-hooks not found\n    at createError (/home/jordan/.nvm/versions/node/v13.12.0/lib/node_modules/bower/lib/util/createError.js:4:15)\n    at /home/jordan/.nvm/versions/node/v13.12.0/lib/node_modules/bower/lib/core/resolverFactory.js:243:23\n    at _fulfilled (/home/jordan/.nvm/versions/node/v13.12.0/lib/node_modules/bower/lib/node_modules/q/q.js:854:54)\n    at /home/jordan/.nvm/versions/node/v13.12.0/lib/node_modules/bower/lib/node_modules/q/q.js:883:30\n    at Promise.promise.promiseDispatch (/home/jordan/.nvm/versions/node/v13.12.0/lib/node_modules/bower/lib/node_modules/q/q.js:816:13)\n    at /home/jordan/.nvm/versions/node/v13.12.0/lib/node_modules/bower/lib/node_modules/q/q.js:624:44\n    at runSingle (/home/jordan/.nvm/versions/node/v13.12.0/lib/node_modules/bower/lib/node_modules/q/q.js:137:13)\n    at flush (/home/jordan/.nvm/versions/node/v13.12.0/lib/node_modules/bower/lib/node_modules/q/q.js:125:13)\n    at processTicksAndRejections (internal/process/task_queues.js:79:11)"
}]
```

thomashoneyman replied:
Oh, shoot. Hmm. What if we add an explicit GitHub URL for these instead of pulling from the package set? We can't publish the new library to Pursuit without a valid Bower file. In the worst case we can just ensure that the Spago.dhall dependencies match the Bower dependencies, at the same major verisons, before release.

JordanMartinez replied:
I mean... there's always [Starsuit](https://spacchetti.github.io/starsuit/)

>  In the worst case we can just ensure that the Spago.dhall dependencies match the Bower dependencies, at the same major verisons, before release.

Meaning the `bower.json` file matches the depdencies' versions specified in the package set?

thomashoneyman replied:
Sorry, that's right.

JordanMartinez replied:
Done.

JordanMartinez committed:
Update halogen-hooks to v0.4.0 and halogen-hooks-extra to v0.6.0

JordanMartinez committed:
Get code to compile again

JordanMartinez committed:
Use newtype: remove need to add `props` to function type signature

JordanMartinez committed:
Update examples to compile

JordanMartinez committed:
Rename Event to SelectEvent; use consistent naming scheme

JordanMartinez committed:
Split spago.dhall file up so that correct dependencies are in correct spot

JordanMartinez committed:
Add license in bower.json file into spago.dhall file; add repository

`spago init` didn't copy the license over

thomashoneyman commented on src/Select.purs:
```diff
@@ -90,266 +77,300 @@ derive instance ordVisibility :: Ord Visibility
+-- | distinguishes one event type from another.
+data SelectEvent
+  = NewSearch String
+  | VisibilityChangedTo Boolean
```
Can we keep this as `VisibilityChangedTo Visibility`?

JordanMartinez replied:
Yup!

JordanMartinez replied:
Fixed in latest commits.

thomashoneyman commented on src/Select.purs:
```diff
@@ -90,266 +77,300 @@ derive instance ordVisibility :: Ord Visibility
+       . Array (HP.IProp (onMouseDown :: ME.MouseEvent | containerProps) (HookM m Unit))
+  , inputProps
+      :: forall inputProps
+       . Array (HP.IProp (InputProps inputProps) (HookM m Unit))
```
These were originally functions `Array _ -> Array _` so that `append` could be pre-applied. That was done because we want to make sure these props get overridden by the user's props, if they supply a duplicate. It's also a little more convenient at the call site -- `props <> [ ]` or `append props []` vs. `setProps []`.

What do you think of returning to that approach? Did you have a reason to move to this one instead?

JordanMartinez replied:
Personally, I thought `props` is better than `setProps []`. If a user wants to override them, wouldn't `props <> [ override1Prop ]` work, too?

thomashoneyman replied:
It does work, it’s just that it’s really easy to write it the other way around and not know why your override isn’t working

JordanMartinez replied:
Ah.... that makes sense. And if you, for whatever reason, decided to combine multiple libraries' props together, you'd get
```purescript
-- no overrides
setSelectProps (setFormlesProps (setCustomHookProps []))

-- with overrides
setSelectProps (setFormlesProps (setCustomHookProps 
  [ overrideSelectProp
  , overrideFormlessProp
  , overrideCustomHookProp
  ]))
```
and even the order in which you 'set' the props matters because Formless could override Select's corresponding prop.

I'll make the change back to `setProps`. If we want to approach this in some other manner, we can in a future PR.

JordanMartinez replied:
Done in latest commit.

thomashoneyman commented on src/Select.purs:
```diff
@@ -90,266 +77,300 @@ derive instance ordVisibility :: Ord Visibility
+    searchDebouncer <- useDebouncer debounceTime \lastSearchState -> Hooks.do
+      case inputRec.inputType of
+        Text -> do
+          -- modifyState (_ { highlightedIndex = (Just 0) })
```
Should this be uncommented?

JordanMartinez replied:
Nope! I swtiched `modifyState` to `H.modify_` and it took a while to realize I needed `Hooks.modify_` not the other one. I likely commented this out, thinking there was another issue elsewhere and then forgot to uncomment it when I found out what the real problem was.

JordanMartinez replied:
Fixed in latest commits.

thomashoneyman commented:
This is really close. I'm happy to merge it in to `master`, but probably not make a release until the README can be updated to be more useful. Users can try out the new Hooks version by updating their `spago.dhall` to have version = "master"` for the library.

I think we should let the library sit a little longer while updating others like Formless before formally releasing, so that we can continue making tweaks as needed.

JordanMartinez committed:
Change Boolean to Visibility

JordanMartinez committed:
Uncomment code and update it to identifier-approach

JordanMartinez commented:
> I think we should let the library sit a little longer while updating others like Formless before formally releasing, so that we can continue making tweaks as needed.

Agreed!

JordanMartinez committed:
Convert props from values into functions that take an array

This helps ensure that any overrides that a user adds actually work
props <> overrides == good
overrides <> props == wrong!

setProps overrides == good (and developer can't screw this up)

JordanMartinez committed:
Update examples to use `set*Props`

thomashoneyman commented on examples/Components/Dropdown.purs:
```diff
@@ -2,24 +2,22 @@ module Components.Dropdown where
-  , selection :: Maybe String
-  , buttonLabel :: String
-  )
+type Slot query = H.Slot query Message
```
This is `Const Void` below for the query type

```suggestion
type Slot = H.Slot (Const Void) Message
```

JordanMartinez replied:
Done.

thomashoneyman commented on examples/Components/Typeahead.purs:
```diff
@@ -44,97 +38,90 @@ data Message
-    , getItemCount: maybe 0 length <<< RD.toMaybe <<< _.available
-    , selections: []
-    , available: RD.NotAsked
+  ( dropdown :: D.Slot (Const Void) Unit )
```
```suggestion
  ( dropdown :: D.Slot Unit )
```

JordanMartinez replied:
Done.

thomashoneyman commented on examples/Components/Typeahead.purs:
```diff
@@ -44,97 +38,90 @@ data Message
-
-  render :: S.State State -> H.ComponentHTML (S.Action Action) ChildSlots Aff
-  render st =
+       selections' <- Hooks.get selectionsId
```
It's OK to use `get` here, but it's not necessary. This could also just be `pure $ Just $ reply selections`.

JordanMartinez replied:
Also done.

thomashoneyman commented on src/Select.purs:
```diff
@@ -90,266 +77,304 @@ derive instance ordVisibility :: Ord Visibility
-
-    lastIndex :: State st -> Int
-    lastIndex = (_ - 1) <<< st.getItemCount <<< userState
+      -- | An array of `IProps` with `ToggleProps`. It
```
These comments aren't going to show up in documentation, so we ought to add this documentation elsewhere, perhaps in a table in the README or something. That can be done in a subsequent update, just noting it here.

JordanMartinez replied:
Yeah... I was wondering about that...

JordanMartinez replied:
So... What if we put it on top of the type declarations (e.g. `ToggleProps`).

JordanMartinez replied:
Or maybe on top of `SelectReturn`? Because more people are likely to hover over that...

thomashoneyman replied:
I think over the type declarations makes the most sense

JordanMartinez replied:
Done. I added `ContainerProps` even though it's not really needed so that we can be consistent.

thomashoneyman reviewed (approved):
I'll leave this open for a couple days for @davezuch or other CitizenNet folks to weigh in if they want to, and then we can merge it.

JordanMartinez committed:
Turn query to Const Void

JordanMartinez committed:
Use `selections` in scope rather than getting it

There is no possibility of stale state occurring here

JordanMartinez committed:
Update bower.json to match spago dependencies

thomashoneyman commented on bower.json:
```diff
@@ -31,12 +31,14 @@
-    "purescript-halogen": "^5.0.0-rc.4",
-    "purescript-record": "^2.0.0"
+    "purescript-halogen": "^5.0.0-rc.9",
+    "purescript-halogen-hooks": "v0.4.0",
```
If these aren't in the bower registry and `bower install` would fail, can we point these at GitHub links instead?

JordanMartinez replied:
Yeah. Done.

JordanMartinez committed:
Update bower.json to point to github URLs

JordanMartinez committed:
Remove unused debug package as a dependency

JordanMartinez committed:
Port prop documentation to type alias; add ContainerProps

JordanMartinez commented:
I think I've addressed everything. Phew!
